File: src/native/RadcamMulticamModule.ts
import { requireNativeModule, requireNativeViewManager } from 'expo-modules-core';
import { ComponentType } from 'react';
import { ViewProps } from 'react-native';

export type CameraMode = 'single' | 'dual' | 'multi';

export type RecordingOptions = {
  codec: 'h264' | 'hevc';
  fileName?: string;
};

export type CapabilityReport = {
  supportsMultiCam: boolean;
  availableRearCameraIDs: string[];
  selectedMode: CameraMode;
};

type RadcamNativeModule = {
  getCapabilitiesAsync(): Promise<CapabilityReport>;
  startSessionAsync(): Promise<void>;
  stopSessionAsync(): Promise<void>;
  startRecordingAsync(options: RecordingOptions): Promise<{ filePath: string }>;
  stopRecordingAsync(): Promise<{ filePath: string | null }>;
};

export type RadcamPreviewViewProps = ViewProps & {
  onFrameMetrics?: (event: { nativeEvent: { fps: number; activeCameras: number } }) => void;
};

export const RadcamMulticamModule = requireNativeModule<RadcamNativeModule>('RadcamMulticam');
export const RadcamPreviewView = requireNativeViewManager('RadcamPreviewView') as ComponentType<RadcamPreviewViewProps>;

File: modules/radcam-multicam/ios/RadcamMulticamModule.swift
import ExpoModulesCore

public final class RadcamMulticamModule: Module {
  public func definition() -> ModuleDefinition {
    Name("RadcamMulticam")

    View(RadcamPreviewView.self) {
      Events("onFrameMetrics")
    }

    AsyncFunction("getCapabilitiesAsync") { () -> [String: Any] in
      MultiCamManager.shared.capabilityReport()
    }

    AsyncFunction("startSessionAsync") {
      MultiCamManager.shared.startSession()
    }

    AsyncFunction("stopSessionAsync") {
      MultiCamManager.shared.stopSession()
    }

    AsyncFunction("startRecordingAsync") { (options: [String: Any]) -> [String: String] in
      let codec = options["codec"] as? String ?? "hevc"
      let filePath = try MultiCamManager.shared.startRecording(codec: codec)
      return ["filePath": filePath]
    }

    AsyncFunction("stopRecordingAsync") { () async -> [String: String?] in
      let filePath = await MultiCamManager.shared.stopRecording()
      return ["filePath": filePath]
    }
  }
}
